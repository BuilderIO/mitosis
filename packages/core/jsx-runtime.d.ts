export * from './src/jsx-types';
