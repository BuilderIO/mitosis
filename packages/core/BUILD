load("@npm//@bazel/typescript:index.bzl", "ts_project")

package(default_visibility = ["//visibility:public"])

ts_project(
    name = "core",
    srcs = glob(
        [
            "src/**/*.ts",
            "src/**/*.tsx",
        ],
        exclude = [
            "**/__tests__/**",
            "**/*.test.ts",
        ],
    ),
    declaration = True,
    preserve_jsx = True,
    # supports_workers = True, # TODO(misko): unclear why enabling this causes tsc to write to source folder
    tsc = "@npm_core//typescript/bin:tsc", # TODO(misko): unclear why this is needed here, but not in other places.
    tsconfig = "//:tsconfig.bazel.json",
    deps = [
        # "@npm_core//:node_modules",
        "@npm_core//acorn",
        "@npm_core//dedent",
        "@npm_core//prettier",
        "@npm_core//lodash",
        "@npm_core//@types/node",
        "@npm_core//@types/babel__core",
        "@npm_core//@types/json5",
        "@npm_core//@types/dedent",
        "@npm_core//@types/prettier",
        "@npm_core//@types/lodash",
        "@npm_core//@types/object-hash",
        "@npm_core//@types/traverse",
        "@npm_core//@types/rollup__plugin-virtual",
        "@npm_core//@builder.io/sdk",
        "@npm_core//@babel/core",
        "@npm_core//@babel/generator",
        "@npm_core//@rollup/plugin-virtual",
        "@npm_core//@angular/compiler",
        "@npm_core//json5",
        "@npm_core//traverse",
        "@npm_core//rollup",
        "@npm_core//csstype",
        "@npm_core//object-hash",
        "@npm_core//axios",
        "@npm_core//liquidjs",
        "@npm_core//vue-template-compiler",
    ],
)
