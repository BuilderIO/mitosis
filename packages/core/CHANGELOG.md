# Change Log

## 0.0.146

### Patch Changes

- 35becd6: Fix: remove all empty text nodes in JSX parser
- f64d9b0: fix: Vue composition API watch deps
- 35becd6: fix: remove redundant {' '} in React generator

## 0.0.145

### Patch Changes

- 9720fb6: Fix: Svelte reactivity in onUpdate dependencies

## 0.0.144

### Patch Changes

- e4f0019: [fix] `preventDefault` in Qwik

## 0.0.142

### Patch Changes

- a4f8311: moved Vue \_classStringToObject logic behind an option

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.
