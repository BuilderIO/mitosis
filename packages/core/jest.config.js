module.exports = {
  clearMocks: true,
  coverageDirectory: 'coverage',
  coverageReporters: ['text', 'clover'],
  coverageThreshold: {
    global: {
      branches: 80,
      functions: 80,
      lines: 80,
      statements: 80,
    },
  },
  globals: {
    'ts-jest': {
      extends: './babel.config.js',
      // module: 'commonjs',
    },
  },
  moduleFileExtensions: ['ts', 'tsx', 'js'],
  modulePathIgnorePatterns: ['dist'],
  moduleNameMapper: {
    '@builder.io/mitosis': '<rootDir>',
  },
  notify: true,
  notifyMode: 'always',
  roots: ['<rootDir>'],
  testMatch: ['**/__tests__/*.test.+(ts|tsx)', '**/*.test.+(ts|tsx)'],
  transform: {
    '^.+\\.raw\\.(ts|tsx|js|jsx)$': 'jest-raw-loader',
    '^.+\\.lite\\.(ts|tsx|js|jsx)$': 'jest-raw-loader',
    '^.+\\.(ts|tsx)$': 'ts-jest',
  },
};
