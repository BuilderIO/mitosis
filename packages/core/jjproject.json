{
  "name": "@builder.io/mitosis",
  "targets": {
    "test": {
      "executor": "@nrwl/vite:vitest",
      "options": {
        "project": "@builder.io/mitosis"
      }
    }
  }
}
