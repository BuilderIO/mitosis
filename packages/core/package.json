{
  "name": "@jsx-lite/core",
  "description": "Write components once, run everywhere. Compiles to Vue, React, Solid, and Liquid. Import code from Figma and Builder.io",
  "keywords": [
    "jsx",
    "compiler",
    "no-code",
    "builder.io",
    "figma",
    "react",
    "vue",
    "solid",
    "liquid",
    "shopify",
    "builder",
    "wysiwyg",
    "drag-drop",
    "jsx-lite",
    "angular"
  ],
  "author": {
    "name": "Builder.io",
    "url": "https://www.builder.io"
  },
  "version": "0.0.3",
  "homepage": "https://github.com/BuilderIO/jsx-lite",
  "main": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "https://github.com/BuilderIO/jsx-lite"
  },
  "scripts": {
    "test": "jest",
    "build": "tsc",
    "release:dev": "npm run build && npm version prerelease --no-git-tag-version && ALLOW_PUBLISH=true npm publish --tag dev",
    "release:patch": "npm run build && npm version patch --no-git-tag-version && ALLOW_PUBLISH=true npm publish",
    "start": "npm run build -- --watch",
    "ts-node": "ts-node -O '{\"module\": \"commonjs\"}'"
  },
  "files": [
    "dist"
  ],
  "dependencies": {
    "@babel/core": "^7.12.3",
    "@babel/plugin-syntax-jsx": "^7.12.1",
    "@babel/preset-typescript": "^7.12.1",
    "axios": "^0.21.0",
    "axios-cache-adapter": "^2.5.0",
    "chalk": "^4.1.0",
    "csstype": "^3.0.4",
    "dedent": "^0.7.0",
    "json5": "^2.1.3",
    "liquidjs": "^9.6.0",
    "lodash": "^4.17.20",
    "object-hash": "^2.0.3",
    "prettier": "^2.1.2",
    "traverse": "^0.6.6",
    "ts-node": "^9.0.0",
    "typescript": "^4.0.5",
    "vue-template-compiler": "^2.6.12"
  },
  "devDependencies": {
    "@builder.io/sdk": "^1.1.20",
    "@jsx-lite/eslint-plugin": "../eslint-plugin",
    "@testing-library/jest-dom": "^5.11.5",
    "@types/babel__core": "^7.1.12",
    "@types/dedent": "^0.7.0",
    "@types/jest": "^26.0.15",
    "@types/json5": "0.0.30",
    "@types/lodash": "^4.14.165",
    "@types/object-hash": "^1.3.4",
    "@types/prettier": "^2.1.5",
    "@types/traverse": "^0.6.32",
    "jest": "^26.6.3",
    "jest-raw-loader": "^1.0.1",
    "ts-jest": "^26.4.4"
  }
}
