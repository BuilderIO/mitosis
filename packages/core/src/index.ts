export * from './flow';

export type Context<T> = {};

// These compile away
export const useState = <T>(obj: T) => obj;
export const useRef = () => null as any;
export const useContext = <T = { [key: string]: any }>(key: Context<T>): T =>
  null as unknown as T;
export const createContext = <T = { [key: string]: any }>(
  value: T,
): Context<T> => null as unknown as Context<T>;
export const setContext = <T = { [key: string]: any }>(
  key: Context<T>,
  value: Partial<T>,
): void => {};
export const onMount = (fn: () => any) => null as any;
export const onUpdate = (fn: () => any, deps?: any[]) => null as any;
export const onCreate = (fn: () => any) => null as any;
export const onInit = (fn: () => any) => null as any;
export const onUnMount = (fn: () => any) => null as any;
export const useDynamicTag = (fn: () => any) => null as any;
export const onError = (fn: () => any) => null as any;
export const useMetadata = (obj: object) => null;

export * from './parsers/jsx';
export * from './parsers/builder';
export * from './parsers/angular';
export * from './parsers/liquid';
export * from './parsers/context';
export * from './generators/vue';
export * from './generators/angular';
export * from './generators/context/react';
export * from './generators/context/vue';
export * from './generators/react';
export * from './generators/solid';
export * from './generators/liquid';
export * from './generators/builder';
export * from './generators/qwik/index';
export * from './symbols/symbol-processor';
export * from './generators/html';
export * from './generators/svelte';
export * from './generators/stencil';
export * from './generators/mitosis';
export * from './generators/template';
export * from './generators/swift-ui';
export * from './generators/react-native';
export * from './helpers/parse-reactive-script';
export * from './helpers/is-mitosis-node';

export * from './types/mitosis-node';
export * from './types/mitosis-component';
export * from './types/config';

export * from './plugins/compile-away-builder-components';
export * from './plugins/compile-away-components';
export * from './plugins/map-styles';

export * from './targets';
