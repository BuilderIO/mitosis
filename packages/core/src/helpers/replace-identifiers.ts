import { types } from '@babel/core';
import generate from '@babel/generator';
import { pipe } from 'fp-ts/lib/function';
import { babelTransformExpression } from './babel-transform';

/**
 * Type hack.
 *
 * We want to augment the `BaseNode` interface to include a `_builder_meta` property but couldn't get
 * `yarn patch-package` to cooperate with us. So we're doing it this way.
 */
type AllowMeta<T = types.Node> = T & {
  _builder_meta?: {
    newlyGenerated: boolean;
  };
};

export type ReplaceTo =
  | string
  | ((accessedProperty: string, matchedIdentifier: string) => string)
  | null;

type ReplaceArgs = {
  code: string;
  from: string | string[];
  to: ReplaceTo;
};

/**
 * Given a `to` function given by the user, figure out the best argument to provide to the `to` function.
 * This function makes a best guess based on the AST structure it's dealing with.
 */
const getToParam = (
  path: babel.NodePath<types.Identifier | types.MemberExpression | types.OptionalMemberExpression>,
): string => {
  if (types.isMemberExpression(path.node) || types.isOptionalMemberExpression(path.node)) {
    if (types.isIdentifier(path.node.property)) {
      // if simple member expression e.g. `props.foo`, returns `foo`
      const propertyName = path.node.property.name;
      return propertyName;
    } else {
      // if nested member expression e.g. `props.foo.bar.baz`, returns `foo.bar.baz`
      const x = generate(path.node.property).code;
      return x;
    }
  } else {
    // if naked identifier e.g. `foo`, returns `foo`
    const nodeName = path.node.name;
    return nodeName;
  }
};

const _replaceIdentifiers = (
  path: babel.NodePath<types.MemberExpression | types.OptionalMemberExpression | types.Identifier>,
  { from, to }: Pick<ReplaceArgs, 'from' | 'to'>,
) => {
  const memberExpressionObject = types.isIdentifier(path.node) ? path.node : path.node.object;
  const normalizedFrom = Array.isArray(from) ? from : [from];

  if (
    !types.isIdentifier(memberExpressionObject) ||
    (path.node as AllowMeta)?._builder_meta?.newlyGenerated
  ) {
    return;
  }

  const matchesFrom = normalizedFrom.includes(memberExpressionObject.name);

  if (matchesFrom) {
    if (to) {
      // `props.foo` to `state`, e.g. `state.foo`
      if (typeof to === 'string') {
        const cleanedIdentifier = pipe(
          // Remove trailing `.` if it exists in the user-provided string, as the dot is generated
          // by babel from the AST
          to.endsWith('.') ? to.substring(0, to.length - 1) : to,
          types.identifier,
        );

        if (types.isIdentifier(path.node)) {
          path.replaceWith(cleanedIdentifier);
        } else {
          path.replaceWith(types.memberExpression(cleanedIdentifier, path.node.property));
        }

        // `props.foo` to (name) => `state.${name}.bar`, e.g. `state.foo.bar`
      } else {
        try {
          const newMemberExpression = pipe(
            getToParam(path),
            (x) => to(x, memberExpressionObject.name),
            (expression) => {
              const [head, ...tail] = expression.split('.');
              return [head, tail.join('.')];
            },
            ([obj, prop]) => {
              const objIdentifier = types.identifier(obj);
              if (prop === '') {
                return objIdentifier;
              } else {
                return types.memberExpression(objIdentifier, types.identifier(prop));
              }
            },
          );

          /**
           * If both `path` and `newMemberExpression` are equal nodes, do nothing.
           * This is to prevent infinite loops when the user-provided `to` function returns the same identifier.
           *
           * The infinite loop probably happens because we end up traversing the new `Identifier` node again?
           */
          if (generate(path.node).code === generate(newMemberExpression).code) {
            return;
          }
          (newMemberExpression as AllowMeta)._builder_meta = { newlyGenerated: true };
          path.replaceWith(newMemberExpression);
        } catch (err) {
          console.debug('Could not replace node.');
          // throw err;
        }
      }
    } else {
      if (types.isIdentifier(path.node)) {
        console.debug(`Could not replace Identifier with nothing.`);
      } else {
        // if we're looking at a member expression, e.g. `props.foo` and no `to` was provided, then we want to strip out
        // the identifier and end up with `foo`. So we replace the member expression with just its `property` value.
        path.replaceWith(path.node.property);
      }
    }
  }
};

/**
 * @deprecated Use `replaceNodes` instead.
 */
export const replaceIdentifiers = ({ code, from, to }: ReplaceArgs) => {
  try {
    return pipe(
      babelTransformExpression(code, {
        MemberExpression(path) {
          _replaceIdentifiers(path, { from, to });
        },
        OptionalMemberExpression(path) {
          _replaceIdentifiers(path, { from, to });
        },
        Identifier(path) {
          // we only want to ignore certain identifiers:
          if (
            // (optional) member expressions are already handled in other visitors
            !types.isMemberExpression(path.parent) &&
            !types.isOptionalMemberExpression(path.parent) &&
            // function declaration identifiers shouldn't be transformed
            !types.isFunctionDeclaration(path.parent)
            // variable declaration identifiers shouldn't be transformed
            // !(types.isVariableDeclarator(path.parent) && path.parent.id === path.node)
          ) {
            _replaceIdentifiers(path, { from, to });
          }
        },
      }),
      // merely running `babel.transform` will add spaces around the code, even if we don't end up replacing anything.
      // we have some other code downstream that cannot have untrimmed spaces, so we need to trim the output.
      (code) => code.trim(),
    );
  } catch (err) {
    throw err;
  }
};

export const replaceStateIdentifier = (to: ReplaceArgs['to']) => (code: string) =>
  replaceIdentifiers({ code, from: 'state', to });

export const replacePropsIdentifier = (to: ReplaceArgs['to']) => (code: string) =>
  replaceIdentifiers({ code, from: 'props', to });

const isNewlyGenerated = (node: types.Node) => (node as AllowMeta)?._builder_meta?.newlyGenerated;

/**
 * Replaces all instances of a Babel AST Node with a new Node within a code string.
 * Uses `generate()` to convert the Node to a string and compare them.
 */
export const replaceNodes = ({
  code,
  nodeMaps,
}: {
  code: string;
  nodeMaps: {
    from: types.Node;
    condition?: (path: babel.NodePath<types.Node>) => boolean;
    to: types.Node;
  }[];
}) => {
  const searchAndReplace = (path: babel.NodePath<types.Node>) => {
    if (isNewlyGenerated(path.node) || isNewlyGenerated(path.parent)) return;

    for (const { from, to, condition } of nodeMaps) {
      if (isNewlyGenerated(path.node) || isNewlyGenerated(path.parent)) return;
      // if (path.node.type !== from.type) return;

      const matchesCondition = condition ? condition(path) : true;

      if (generate(path.node).code === generate(from).code && matchesCondition) {
        const x = types.cloneNode(to);
        (x as AllowMeta)._builder_meta = { newlyGenerated: true };
        try {
          path.replaceWith(x);
        } catch (err) {
          console.log('error replacing', {
            code,
            orig: generate(path.node).code,
            to: generate(x).code,
          });
          // throw err;
        }
      }
    }
  };

  return babelTransformExpression(code, {
    MemberExpression(path) {
      searchAndReplace(path);
    },
    Identifier(path) {
      searchAndReplace(path);
    },
    OptionalMemberExpression(path) {
      searchAndReplace(path);
    },
  });
};
