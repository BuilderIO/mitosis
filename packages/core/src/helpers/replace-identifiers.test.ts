import { types } from '@babel/core';
import { replaceIdentifiers, replaceNodes } from './replace-identifiers';

type Spec = Parameters<typeof replaceIdentifiers>[0];

const CODE_BLOCK = `
function updateThing() {
  state.thing1 = props.thing2 + 123;

  state?.fn(props?.abc.foo);

  const x = someRandomObj.state.foo
  const y = someRandomObj.props.state.foo
}
`;

const TEST_SPECS: Spec[] = [
  {
    from: 'props',
    to: '$SPECIAL',
    code: CODE_BLOCK,
  },
  {
    from: ['props', 'state'],
    to: null,
    code: CODE_BLOCK,
  },
  {
    from: ['props', 'state'],
    to: 'this',
    code: CODE_BLOCK,
  },
  {
    code: '!state.useLazyLoading() || load',
    from: ['scrollListener', 'imageLoaded', 'setLoaded', 'useLazyLoading', 'isBrowser', 'load'],
    to: (name) => `state.${name}`,
  },
  {
    code: `state.name = 'PatrickJS onInit' + props.hi;`,
    from: ['props'],
    to: (name) => `this.${name}`,
  },
  {
    code: 'state.lowerCaseName()',
    from: 'state',
    to: (name) => (name === 'children' ? '$$slots.default' : name),
  },
  {
    code: `
    const x = {
      foo: bar,
      test: 123,
    }

    const foo = x.foo;

    const y = {
      l: x.foo,
      m: foo
    }

    const bar = foo;
    `,
    from: ['foo', 'test'],
    to: (name) => `${name}.value`,
  },
];

describe('replaceIdentifiers', () => {
  TEST_SPECS.forEach((args, index) => {
    test(`Check #${index}`, () => {
      const output = replaceIdentifiers(args);
      expect(output).toMatchSnapshot();
    });
  });
});

describe('newReplacer', () => {
  test('Check #1', () => {
    const code = `
  const [childrenContext] = useState(
    useTarget({
      reactNative: {
        apiKey: props.context.value.apiKey,
        apiVersion: props.context.value.apiVersion,
        localState: props.context.value.localState,
        rootState: props.context.value.rootState,
        rootSetState: props.context.value.rootSetState,
        content: props.context.value.content,
        context: props.context.value.context,
        registeredComponents: props.context.value.registeredComponents,
        inheritedStyles: extractTextStyles(
          getReactNativeBlockStyles({
            block: state.useBlock,
            context: props.context.value,
            blockStyles: state.attributes.style,
          })
        ),
      },
      default: props.context.value,
    }),
    { reactive: true }
  );
`;
    const thing = types.memberExpression(
      types.memberExpression(types.identifier('props'), types.identifier('context')),
      types.identifier('value'),
    );

    const to = types.memberExpression(types.identifier('props'), types.identifier('$context'));
    const output = replaceNodes({ code, nodeMaps: [{ from: thing, to }] });
    expect(output).toMatchSnapshot();
  });
});
