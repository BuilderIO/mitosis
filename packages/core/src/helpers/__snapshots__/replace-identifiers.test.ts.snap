// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`newReplacer > Check #1 1`] = `
"const [childrenContext] = useState(useTarget({
  reactNative: {
    apiKey: props.$context.apiKey,
    apiVersion: props.$context.apiVersion,
    localState: props.$context.localState,
    rootState: props.$context.rootState,
    rootSetState: props.$context.rootSetState,
    content: props.$context.content,
    context: props.$context.context,
    registeredComponents: props.$context.registeredComponents,
    inheritedStyles: extractTextStyles(getReactNativeBlockStyles({
      block: state.useBlock,
      context: props.$context,
      blockStyles: state.attributes.style
    }))
  },
  default: props.$context
}), {
  reactive: true
})"
`;

exports[`replaceIdentifiers > Check #0 1`] = `
"function updateThing() {
  state.thing1 = $SPECIAL.thing2 + 123;
  state?.fn($SPECIAL.abc.foo);
  const x = someRandomObj.state.foo;
  const y = someRandomObj.props.state.foo;
}"
`;

exports[`replaceIdentifiers > Check #1 1`] = `
"function updateThing() {
  thing1 = thing2 + 123;
  fn(abc.foo);
  const x = someRandomObj.state.foo;
  const y = someRandomObj.props.state.foo;
}"
`;

exports[`replaceIdentifiers > Check #2 1`] = `
"function updateThing() {
  this.thing1 = this.thing2 + 123;
  this.fn(this.abc.foo);
  const x = someRandomObj.state.foo;
  const y = someRandomObj.props.state.foo;
}"
`;

exports[`replaceIdentifiers > Check #3 1`] = `"!state.useLazyLoading() || state.load"`;

exports[`replaceIdentifiers > Check #4 1`] = `"state.name = 'PatrickJS onInit' + this.hi"`;

exports[`replaceIdentifiers > Check #5 1`] = `"lowerCaseName()"`;

exports[`replaceIdentifiers > Check #6 1`] = `
"const x = {
  foo: bar,
  test: 123
};
const foo.value = x.foo;
const y = {
  l: x.foo,
  m: foo
};
const bar = foo.value"
`;
